### PR TITLE
Move mock version of dgl_sparse library to DGL main repo

### PR DESCRIPTION
**Pull request by isratnisa**
## Description
Resolves issue https://github.com/dmlc/dgl/issues/4489
This PR is  a duplicate of [PR #4500](https://github.com/dmlc/dgl/pull/4500). The old PR was created on the base repo `dmlc/dgl` (not on a forked repo) by mistake. 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: tests/mock_sparse/test_elementwise_op_sp.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+@parametrize_idtype
+@parametrize_dtype
+@pytest.mark.parametrize('v_scalar', [2, 2.5])
+def test_scalar_op_sparse(idtype, dtype, v_scalar):
+    row = torch.randint(1, 500, (100,))
+    col = torch.randint(1, 500, (100,))
+    val = torch.rand(100)
+    A = SparseMatrix(row, col, val)
+    out_torch = v_scalar * A.val
+    out_spMat = v_scalar * A
```
Comment (jermainewang) [inline tests/mock_sparse/test_elementwise_op_sp.py:69]: Missing `A / v_scalar`

**Review by jermainewang — CHANGES_REQUESTED** [review]
Some common issues:
* Please make sure examples in the docstring are complete and their outputs are accurate.
* Please directly print out sparse matrix as much as possible.
* Error messages are not general enough which may confuse users in some cases.

**Comment by jermainewang** [comment]
The unit tests for sparse matrix class are missing in the PR. @isratnisa  @mufeili will you guys do it in another PR?

File: tests/mock_sparse/test_elementwise_op_sp.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ops = {
+    "+": operator.add,
+    "-": operator.sub,
+    "*": operator.mul,
+    "/": operator.truediv,
+}
+
+@parametrize_idtype
+@parametrize_dtype
+def test_sparse_op_sparse(idtype, dtype):
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tests/mock_sparse/test_elementwise_op_sp.py]: Done.

File: tests/mock_sparse/test_elementwise_op_sp.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    row = torch.randint(1, 500, (100,))
+    col = torch.randint(1, 500, (100,))
+    val = torch.rand(100)
+    A = SparseMatrix(row, col, val)
+
+    def _test(A, op):
+        out_torch = eval("A.val " + op + " A.val")
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply tests/mock_sparse/test_elementwise_op_sp.py]: Done.

File: python/dgl/mock_sparse/elementwise_op_sp.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    >>> C = A - A
+    >>> C.val
+    tensor([0, 0, 0])
+    >>> C = A(w2) - A(w1)
+    tensor([-4, -7, -8])
+    """
+    if isinstance(A, SparseMatrix) and isinstance(B, SparseMatrix):
+        assert A.shape == B.shape, 'The shape of sparse matrix A {} and' \
+        'B {} must match'.format(A.shape, B.shape)
+        check_sparsity(A, B)
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/dgl/mock_sparse/elementwise_op_sp.py]: Added support to add two matrices of different sparsity.

File: docs/source/api/python/dgl.sparse_v0.rst [hunk 5]
```diff
@@ -0,0 +1,14 @@
+.. _apibackend:
+
+dgl.sparse
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply docs/source/api/python/dgl.sparse_v0.rst:3]: Do we want to say `mock_sparse` in the title? The warning section I think takes care of that.

File: python/dgl/mock_sparse/elementwise_op_sp.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
+"""dgl elementwise operators for sparse matrix module."""
+import torch
+from .sp_matrix import SparseMatrix
+
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/dgl/mock_sparse/elementwise_op_sp.py:4]: Do we want to use `import *` from `sp_matrix.py`?

File: docs/source/api/python/dgl.sparse_v0.rst [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+=================================
+
+`dgl_sparse` is a library for sparse operators that are commonly used in GNN models.
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply docs/source/api/python/dgl.sparse_v0.rst:6]: Done.

File: python/dgl/mock_sparse/sp_matrix.py [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+
+    Examples
+    --------
+
+    Case1: Sparse matrix with row and column indices without values.
+
+    >>> src = torch.tensor([1, 1, 2])
+    >>> dst = torch.tensor([2, 4, 3])
+    >>> A = create_from_coo(src, dst)
+    >>> A.shape
```
[the remark this answers is not in the document]
Comment (isratnisa) [reply python/dgl/mock_sparse/sp_matrix.py]: Done.

File: python/dgl/mock_sparse/elementwise_op_sp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/mock_sparse/elementwise_op_sp.py:4]: Yes, you should add `from .xxx import *` in the package `__init__.py` so users can use `dgl.mock_sparse.some_api`.

File: tests/pytorch/mock_sparse/test_elementwise_op_sp.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
+import numpy as np
+import pytest
+import dgl
+import dgl.backend as F
+import torch
+import numpy
+import operator
+from dgl.mock_sparse import elementwise_op_sp
+from dgl.mock_sparse.sp_matrix import SparseMatrix
```
Comment (jermainewang) [inline tests/pytorch/mock_sparse/test_elementwise_op_sp.py:9]: Users shouldn't import from `dgl.mock_sparse.elementwise_op_sp` or `dgl.mock_sparse.sp_matrix`. They should directly import `dgl.mock_sparse.some_api` or `dgl.mock_sparse.SparseMatrix`. This is an issue related to the `__all__` comment.

File: tests/pytorch/mock_sparse/test_elementwise_op_sp.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+def all_close_sparse(A, B):
+    assert torch.allclose(A.indices(), B.indices())
+    assert torch.allclose(A.values(), B.values())
+    assert A.shape == B.shape
+
+@parametrize_idtype
+@parametrize_dtype
+@pytest.mark.parametrize('op', [operator.add, operator.sub, operator.mul, operator.truediv])
+def test_sparse_op_sparse(idtype, dtype, op):
```
Comment (jermainewang) [inline tests/pytorch/mock_sparse/test_elementwise_op_sp.py:21]: Please add a similar test for `dgl.mock_sparse.add(A, B)`, etc. This is to verify that the APIs are correctly exposed.

**Review by jermainewang — CHANGES_REQUESTED** [review]
Just one last suggestion about the API namespace. Otherwise, LGTM. Will approve afterwards.

**Review by jermainewang — APPROVED** [review]
LGTM

**Pull request merged**